### PR TITLE
Fix JNumber Deserialize

### PR DESCRIPTION
**Pull request by Ashuaidehao**
Close https://github.com/neo-project/neo/issues/2879

### CODE REVIEW

**Review by Liaojinghui — APPROVED** [review]

**Comment by Liaojinghui** [comment]
May you also add a unit test for that

**Review by Liaojinghui — APPROVED** [review]

File: src/Neo/SmartContract/JsonSerializer.cs [hunk 1]
```diff
@@ -191,8 +191,7 @@ private static StackItem Deserialize(JToken json, ref uint maxStackSize, Referen
                 case JNumber num:
                     {
                         if ((num.Value % 1) != 0) throw new FormatException("Decimal value is not allowed");
-
-                        return (BigInteger)num.Value;
+                        return BigInteger.Parse(num.Value.ToString(), System.Globalization.NumberStyles.Float);
```
Comment (AnnaShaleva) [inline src/Neo/SmartContract/JsonSerializer.cs:194]: That should be enough to fit integers that are allowed by our 256-bit VM. I've checked this fix against the neo-go node, and now both nodes are able to handle big numbers correctly.

File: src/Neo/SmartContract/JsonSerializer.cs [hunk 2]
```diff
@@ -191,8 +191,7 @@ private static StackItem Deserialize(JToken json, ref uint maxStackSize, Referen
                 case JNumber num:
                     {
                         if ((num.Value % 1) != 0) throw new FormatException("Decimal value is not allowed");
```
Comment (AnnaShaleva) [inline src/Neo/SmartContract/JsonSerializer.cs:193]: And BTW, this line saves us from things like `100.123e+2`. The exception is thrown on attempt to parse it, thus, the second problem described in #2879 isn't relevant:
```
Error Message:
   Test method Neo.UnitTests.SmartContract.UT_JsonSerializer.Deserialize_Array_Bool_Str_Num threw exception: 
System.FormatException: Decimal value is not allowed
  Stack Trace:
      at Neo.SmartContract.JsonSerializer.Deserialize(JToken json, UInt32& maxStackSize, ReferenceCounter referenceCounter) in /home/anna/Documents/GitProjects/neo-project/neo/src/Neo/SmartContract/JsonSerializer.cs:line 193
   at Neo.SmartContract.JsonSerializer.Deserialize(JToken json, UInt32& maxStackSize, ReferenceCounter referenceCounter) in /home/anna/Documents/GitProjects/neo-project/neo/src/Neo/SmartContract/JsonSerializer.cs:line 184
   at Neo.SmartContract.JsonSerializer.Deserialize(JToken json, ExecutionEngineLimits limits, ReferenceCounter referenceCounter) in /home/anna/Documents/GitProjects/neo-project/neo/src/Neo/SmartContract/JsonSerializer.cs:line 168
   at Neo.UnitTests.SmartContract.UT_JsonSerializer.Deserialize_Array_Bool_Str_Num() in /home/anna/Documents/GitProjects/neo-project/neo/tests/Neo.UnitTests/SmartContract/UT_JsonSerializer.cs:line 290
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```

**Review by AnnaShaleva — APPROVED** [review]

**Comment by superboyiii** [comment]
> @superboyiii Do you think a test for consistancy is necessary?

We can use Basilisk

**Comment by roman-khimov** [comment]
4c9e33b29bb75ae2a77e180f82879195fe541afa04aabe0e4a1a1ec858629938. We don't have any doing 1.2345e+3, but the possibility of this being unnoticed with 3.5 makes me suggest that Basilisk is the right HF for this change.

**Comment by superboyiii** [comment]
> 4c9e33b29bb75ae2a77e180f82879195fe541afa04aabe0e4a1a1ec858629938. We don't have any doing 1.2345e+3, but the possibility of this being unnoticed with 3.5 makes me suggest that Basilisk is the right HF for this change.

Yes, let's use Basilisk @Ashuaidehao

File: src/Neo/SmartContract/JsonSerializer.cs (hunk 1 above)
Comment (shargon) [inline src/Neo/SmartContract/JsonSerializer.cs:194]:
```suggestion
                        return BigInteger.Parse(num.Value.ToString(CultureInfo.InvariantCulture), NumberStyles.Float);
```

**Review by shargon — CHANGES_REQUESTED** [review]

**Comment by shargon** [comment]
Ready for test @superboyiii

**Review by AnnaShaleva — APPROVED** [review]

**Review by Liaojinghui — APPROVED** [review]

**Comment by superboyiii** [comment]
> Ready for test @superboyiii

Work as the expected! All data are compatible either on mainnet or  testnet if add a future hard fork.

**Review by superboyiii — APPROVED** [review]

**Comment by superboyiii** [comment]
```
 public static object Print1(string json)
        {
            //input json = "[200.500000E+005,200.500000e+5,1.1234e+10,9.05E+29]";
            var parsed = StdLib.JsonDeserialize(json);
            return parsed;
        }
```
Input: `"[200.500000E+005,200.500000e+5,1.1234e+10,9.05E+29]"`
Result:
```
{
    "jsonrpc": "2.0",
    "id": 3,
    "result": {
        "script": "DDNbMjAwLjUwMDAwMEUrMDA1LDIwMC41MDAwMDBlKzUsMS4xMjM0ZSsxMCw5LjA1RSsyOV0RwB8MBnByaW50MQwURp5xWNKpMah1VXMjSI/rtnNdtX5BYn1bUg==",
        "state": "HALT",
        "gasconsumed": "3646350",
        "exception": null,
        "notifications": [],
        "stack": [
            {
                "type": "Array",
                "value": [
                    {
                        "type": "Integer",
                        "value": "20050000"
                    },
                    {
                        "type": "Integer",
                        "value": "20050000"
                    },
                    {
                        "type": "Integer",
                        "value": "11234000000"
                    },
                    {
                        "type": "Integer",
                        "value": "905000000000000044735708266496"
                    }
                ]
            }
        ]
    }
}
```
value[3] comparing between this PR and previous one:
635658ccbcf8fe5a6762c30bc3ef122
Works as expected.

**Comment by superboyiii** [comment]
@shargon @Liaojinghui merge?

**Comment by Liaojinghui** [comment]
LGTM

**Pull request merged**